File: anxiety-support/backend/bot.js
const { Telegraf } = require('telegraf');
const cron = require('node-cron');
const { db } = require('./database');

let bot;

function miniAppUrl(section = 'home') {
  const base = process.env.APP_BASE_URL;
  if (!base) return null;
  const url = new URL(base);
  url.searchParams.set('section', section);
  return url.toString();
}

function buildMainKeyboard() {
  const webUrl = miniAppUrl('home');
  if (!webUrl) return { reply_markup: { remove_keyboard: true } };

  return {
    reply_markup: {
      inline_keyboard: [
        [{ text: 'Open Anxiety Support Mini App', web_app: { url: webUrl } }],
        [
          { text: 'Breathing', web_app: { url: miniAppUrl('breathe') } },
          { text: 'Grounding', web_app: { url: miniAppUrl('calm') } }
        ],
        [
          { text: 'Journal', web_app: { url: miniAppUrl('journal') } },
          { text: 'Daily Check-In', web_app: { url: miniAppUrl('home') } }
        ]
      ]
    }
  };
}

async function setupBot() {
  const token = process.env.TELEGRAM_BOT_TOKEN;
  if (!token) {
    console.warn('TELEGRAM_BOT_TOKEN missing, bot disabled.');
    return null;
  }

  bot = new Telegraf(token);

  bot.start(async (ctx) => {
    const text = [
      `Hi ${ctx.from.first_name || ''} 🌿`,
      'I am your Anxiety Support companion.',
      'Use the Mini App for check-ins, breathing, grounding, CBT journaling, and progress tracking.'
    ].join('\n');
    await ctx.reply(text, buildMainKeyboard());
  });

  bot.command('tip', async (ctx) => {
    const tips = [
      'Try box breathing: inhale 4, hold 4, exhale 4, hold 4.',
      'Notice and name your worry, then challenge it with evidence.',
      'Use 5-4-3-2-1 grounding to reconnect to the present moment.'
    ];
    const tip = tips[Math.floor(Math.random() * tips.length)];
    await ctx.reply(`Quick tip: ${tip}`);
  });

  const sectionMap = { breathe: 'breathe', grounding: 'calm', journal: 'journal', checkin: 'home' };
  for (const [command, section] of Object.entries(sectionMap)) {
    bot.command(command, async (ctx) => {
      const url = miniAppUrl(section);
      if (!url) { await ctx.reply('Mini App URL is not configured yet.'); return; }
      await ctx.reply('Opening Mini App section:', {
        reply_markup: { inline_keyboard: [[{ text: `Open ${section}`, web_app: { url } }]] }
      });
    });
  }

  try {
    await bot.telegram.setMyCommands([
      { command: 'start', description: 'Open support bot and Mini App' },
      { command: 'tip', description: 'Get a quick anxiety tip' },
      { command: 'breathe', description: 'Open breathing exercises' },
      { command: 'grounding', description: 'Open grounding tools' },
      { command: 'journal', description: 'Open CBT journal prompts' },
      { command: 'checkin', description: 'Open daily check-in' }
    ]);

    if (process.env.APP_BASE_URL) {
      await bot.telegram.setChatMenuButton({
        menu_button: {
          type: 'web_app',
          text: 'Open Mini App',
          web_app: { url: miniAppUrl('home') }
        }
      });
    }
  } catch (err) {
    console.error('Bot command/menu setup failed:', err.message);
  }

  const reminderCron = process.env.REMINDER_CRON || '0 9 * * *';
  const reminderTz = process.env.REMINDER_TIMEZONE || 'UTC';

  cron.schedule(reminderCron, async () => {
    const users = db.prepare('SELECT telegram_id, first_name FROM users').all();
    for (const user of users) {
      try {
        const text = [
          `Hi ${user.first_name || 'there'} 🌿`,
          'Friendly reminder: complete your daily check-in and a 2-minute breathing reset.'
        ].join('\n');
        await bot.telegram.sendMessage(user.telegram_id, text, buildMainKeyboard());
      } catch (e) {
        console.error('Reminder send failed:', e.message);
      }
    }
  }, { timezone: reminderTz });

  return bot;
}

async function setWebhook() {
  if (!bot) return;
  const webhookUrl = process.env.BOT_WEBHOOK_URL;
  const secretToken = process.env.TELEGRAM_WEBHOOK_SECRET;
  if (!webhookUrl) return;

  try {
    await bot.telegram.setWebhook(webhookUrl, secretToken ? { secret_token: secretToken } : {});
    console.log('Webhook configured:', webhookUrl);
  } catch (err) {
    console.error('Failed to set webhook:', err.message);
  }
}

async function handleUpdate(req, res) {
  if (!bot) return res.status(503).send('Bot not configured');

  const expectedSecret = process.env.TELEGRAM_WEBHOOK_SECRET;
  const receivedSecret = req.headers['x-telegram-bot-api-secret-token'];

  if (expectedSecret && expectedSecret !== receivedSecret) {
    return res.status(401).send('Invalid secret token');
  }

  try {
    await bot.handleUpdate(req.body);
    res.sendStatus(200);
  } catch (err) {
    console.error('Webhook update error:', err.message);
    res.sendStatus(500);
  }
}

module.exports = {
  setupBot,
  setWebhook,
  handleUpdate,
};

File: anxiety-support/backend/bot.py
import asyncio
import logging
import os
import random
from datetime import datetime, timedelta, timezone

from aiogram import Bot, Dispatcher, Router
from aiogram.filters import Command
from aiogram.types import (
    BotCommand,
    InlineKeyboardButton,
    InlineKeyboardMarkup,
    MenuButtonWebApp,
    Message,
    Update,
    WebAppInfo,
)
from fastapi import APIRouter, Header, Request
from fastapi.responses import Response

logger = logging.getLogger(__name__)

_bot: Bot | None = None
_dp: Dispatcher | None = None

tg_router = Router()
webhook_router = APIRouter()


def mini_app_url(section: str = "home") -> str:
    base = os.getenv("APP_BASE_URL", "").rstrip("/")
    return f"{base}?section={section}"


def main_keyboard() -> InlineKeyboardMarkup:
    return InlineKeyboardMarkup(
        inline_keyboard=[
            [InlineKeyboardButton(text="🌿 Open Anxiety Support", web_app=WebAppInfo(url=mini_app_url("home")))],
            [
                InlineKeyboardButton(text="🌬 Breathing", web_app=WebAppInfo(url=mini_app_url("breathe"))),
                InlineKeyboardButton(text="🌱 Grounding", web_app=WebAppInfo(url=mini_app_url("calm"))),
            ],
            [
                InlineKeyboardButton(text="📓 Journal", web_app=WebAppInfo(url=mini_app_url("journal"))),
                InlineKeyboardButton(text="✅ Check-In", web_app=WebAppInfo(url=mini_app_url("home"))),
            ],
        ]
    )


@tg_router.message(Command("start"))
async def cmd_start(message: Message):
    name = message.from_user.first_name or ""
    await message.answer(
        f"Hi {name} 🌿\n"
        "I'm your Anxiety Support companion.\n"
        "Use the Mini App for daily check-ins, breathing, grounding, CBT journaling, and progress tracking.",
        reply_markup=main_keyboard(),
    )


@tg_router.message(Command("tip"))
async def cmd_tip(message: Message):
    tips = [
        "Try box breathing: inhale 4s → hold 4s → exhale 4s → hold 4s.",
        "Notice and name your worry, then challenge it with evidence.",
        "Use 5-4-3-2-1 grounding to reconnect to the present moment.",
        "Self-compassion lowers stress. Speak to yourself kindly.",
        "Tiny actions reduce overwhelm — choose one 2-minute task.",
    ]
    await message.answer(f"💡 Quick tip: {random.choice(tips)}")


async def _open_section(message: Message, section: str, label: str):
    kb = InlineKeyboardMarkup(
        inline_keyboard=[[InlineKeyboardButton(text=f"Open {label}", web_app=WebAppInfo(url=mini_app_url(section)))]]
    )
    await message.answer("Opening Mini App:", reply_markup=kb)


@tg_router.message(Command("breathe"))
async def cmd_breathe(message: Message):
    await _open_section(message, "breathe", "Breathing")


@tg_router.message(Command("grounding"))
async def cmd_grounding(message: Message):
    await _open_section(message, "calm", "Grounding")


@tg_router.message(Command("journal"))
async def cmd_journal(message: Message):
    await _open_section(message, "journal", "Journal")


@tg_router.message(Command("checkin"))
async def cmd_checkin(message: Message):
    await _open_section(message, "home", "Check-In")


async def _send_daily_reminders():
    if not _bot:
        return
    from database import get_db
    db = get_db()
    users = db.execute("SELECT telegram_id, first_name FROM users").fetchall()
    for user in users:
        try:
            name = user["first_name"] or "there"
            await _bot.send_message(
                chat_id=user["telegram_id"],
                text=f"Hi {name} 🌿\nFriendly reminder: complete your daily check-in and a 2-minute breathing reset.",
                reply_markup=main_keyboard(),
            )
        except Exception as e:
            logger.warning(f"Reminder failed for {user['telegram_id']}: {e}")


async def _reminder_loop():
    while True:
        now = datetime.now(timezone.utc)
        next_run = now.replace(hour=9, minute=0, second=0, microsecond=0)
        if next_run <= now:
            next_run += timedelta(days=1)
        await asyncio.sleep((next_run - now).total_seconds())
        await _send_daily_reminders()


async def setup_bot():
    global _bot, _dp
    token = os.getenv("TELEGRAM_BOT_TOKEN")
    if not token:
        logger.warning("TELEGRAM_BOT_TOKEN not set — bot disabled")
        return

    _bot = Bot(token=token)
    _dp = Dispatcher()
    _dp.include_router(tg_router)

    try:
        await _bot.set_my_commands([
            BotCommand(command="start",     description="Open support bot and Mini App"),
            BotCommand(command="tip",       description="Get a quick anxiety tip"),
            BotCommand(command="breathe",   description="Open breathing exercises"),
            BotCommand(command="grounding", description="Open grounding tools"),
            BotCommand(command="journal",   description="Open CBT journal prompts"),
            BotCommand(command="checkin",   description="Open daily check-in"),
        ])
        if os.getenv("APP_BASE_URL"):
            await _bot.set_chat_menu_button(
                menu_button=MenuButtonWebApp(
                    text="Open Mini App",
                    web_app=WebAppInfo(url=mini_app_url("home")),
                )
            )
    except Exception as e:
        logger.error(f"Bot config error: {e}")

    asyncio.create_task(_reminder_loop())
    logger.info("Bot ready: @%s", (await _bot.get_me()).username)


async def set_webhook():
    if not _bot:
        return
    url = os.getenv("BOT_WEBHOOK_URL")
    secret = os.getenv("TELEGRAM_WEBHOOK_SECRET")
    if not url:
        return
    await _bot.set_webhook(url, secret_token=secret)
    logger.info("Webhook set: %s", url)


@webhook_router.post("/telegram/webhook")
async def telegram_webhook(
    request: Request,
    x_telegram_bot_api_secret_token: str = Header(default=""),
):
    expected = os.getenv("TELEGRAM_WEBHOOK_SECRET", "")
    if expected and x_telegram_bot_api_secret_token != expected:
        return Response(status_code=401)
    if not _dp or not _bot:
        return Response(status_code=503)
    update = Update.model_validate(await request.json())
    await _dp.feed_update(_bot, update)
    return Response(status_code=200)
